src/pointer/ptr.rs

```diff
@@ -0,0 +1,1010 @@
+        ///   [`ALIGNMENT_INVARIANT`](invariant::Alignment).
+        /// - `ptr` conforms to the validity invariant of
+        ///   [`VALIDITY_INVARIANT`](invariant::Validity).
+        pub(super) const unsafe fn from_ptr<
+            const INITIAL_ALIASING_INVARIANT: invariant::Aliasing,
+            const INITIAL_ALIGNMENT_INVARIANT: invariant::Alignment,
+            const INITIAL_VALIDITY_INVARIANT: invariant::Validity,
+        >(
+            ptr: Ptr<
+                'a,
+                T,
+                INITIAL_ALIASING_INVARIANT,
+                INITIAL_ALIGNMENT_INVARIANT,
+                INITIAL_VALIDITY_INVARIANT,
+            >,
+        ) -> Self {
+            // SAFETY: The caller has promised to satisfy all parameterized
+            // invariants of `Ptr`. `Ptr`'s other invariants are satisfied
+            // by-contract by the source `Ptr`.
+            unsafe { Self::new(ptr.as_non_null()) }
+        }
+
+        /// Converts this `Ptr<T>` to a [`NonNull<T>`].
+        ///
+        /// Note that this method does not consume `self`. Callers should be
+        /// aware that safe code may assume that the aliasing invariants of
+        /// `self` are not violated, and that the `NonNull` may be used in ways
+        /// that violate those assumptions.
+        pub(crate) const fn as_non_null(&self) -> NonNull<T> {
+            self.ptr
+        }
+    }
+}
+
+pub use def::Ptr;
+
+/// The parameterized invariants of a [`Ptr`].
+//
+// These items are named according to the conventions for enums, should be
+// treated as if they are enums (i.e., only referenced by name), and should be
+// re-defined as enums as soon as it is possible to do so.
+#[allow(non_upper_case_globals)]
+#[doc(hidden)]
+pub mod invariant {
+    /// The aliasing invariant of a [`Ptr`][super::Ptr].
+    ///
+    /// Either:
+    /// - [`AnyAliasing`]
+    /// - [`Shared`]
+    /// - [`Unique`]
+    pub type Aliasing = u8;
+
+    /// No aliasing invariant.
+    pub const AnyAliasing: Aliasing = 1 << 0;
+
+    /// The pointer adheres to the aliasing rules of a `&T`.
+    pub const Shared: Aliasing = 1 << 2;
+
+    /// The pointer adheres to the aliasing rules of a `&mut T`.
+    pub const Unique: Aliasing = 1 << 3;
+
+    /// The alignment invariant of a [`Ptr`][super::Ptr].
+    ///
+    /// Either:
+    /// - [`AnyAlignment`]
+    /// - [`Aligned`]
+    pub type Alignment = i8;
+
+    /// The referent is not necessarily aligned.
+    pub const AnyAlignment: Alignment = 1 << 4;
+
+    /// The referent is aligned.
+    pub const Aligned: Alignment = 1 << 5;
+
+    /// The validity invariant of a [`Ptr`][super::Ptr].
+    ///
+    /// Either:
+    /// - [`AnyValidity`]
+    /// - [`AsInitialized`]
+    /// - [`Valid`]
+    pub type Validity = u16;
+
+    /// No validity invariant.
+    pub const AnyValidity: Validity = 1 << 6;
+
+    /// The byte ranges initialized in `T` are also initialized in `Ptr`'s
+    /// referent.
```

Comment:
I don't believe this description is sufficiently precise, especially when it comes to enums (where the required initialization state of a given byte is dynamic, and depends upon preceding enum discriminants). IIUC, this basically needs to contain the entire text of the old safety precondition from `is_bit_valid`.

Reply:
Update: Just saw that you moved the safety comment to `assume_as_initialized`. Maybe we can move those here and have `assume_as_initialized` reference these docs?

[TODO for self: Go through and audit uses to make sure safety comments cite this safety precondition and justify it]